magento/pwa-studio
# [feature]: Disable cart mutations while others are in flight

## Description

~When any mutation occurs on the cart page we need to disable other inputs/mutations otherwise it is very possible (and likely) that the cache data gets out of sync with the server.~

When any mutation occurs, queue it to avoid a race condition where cart data gets out of sync with the server. While any mutation is in flight (ie `called && loading`), hide the price summary values until complete.

Instead of applying a mask to basically enforce that a user cannot make a change while another is in flight we can use an Apollo [link](https://www.apollographql.com/docs/link/stateful/) to prevent mutations firing at the same time, regardless of how many are "queued". I created one specifically for this purpose, [apollo-link-mutation-queue](https://github.com/adobe/apollo-link-mutation-queue).

The idea is simple - if the link receives a mutation and there are no other mutations in flight, execute it immediately. If the link receives a mutation and there is a mutation already in flight, enqueue it to be executed after the in-flight mutation resolves.

If we ever need to allow a mutation to skip the queue, just provide `{ context: { skipQueue: true }` and it will be requested in parallel with any other requests.

## Related Issue





Closes [PWA-385](https://jira.corp.magento.com/browse/PWA-385).

## Acceptance 



### Verification Stakeholders

### Specification



### Verification Steps


1. Add a product and go to `/cart`.
2. Make a mutation such as updating quantity, applying coupon/gift card etc. The prices should disappear while the mutation is in flight.

## Screenshots / Screen Captures (if appropriate)
[![Image from Gyazo](https://i.gyazo.com/e8f647efb00b0471d5b86d50e8a88c93.gif)](https://gyazo.com/e8f647efb00b0471d5b86d50e8a88c93)

## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.
